packages/main/src/Tree.js

```diff
@@ -0,0 +1,378 @@
+		this._observer.observe(this, { attributes: true, childList: true, subtree: true });
+	}
+
+	onExitDOM() {
+		this._observer.disconnect();
+	}
+
+	onTreeStructureChange() {
+		// setTimeout is needed for IE11 so that it does not interfere with ItemNavigation.js and its await on RenderScheduler.
+		// Otherwise this invalidation happens too soon and the ItemNavigation is blocked on waiting the tree to finish
+		setTimeout(() => {
+			this._listItems = []; // trigger onBeforeRendering by modifying the tracked property and force tree re-build
+		}, 0);
+	}
+
+	get list() {
+		return this.shadowRoot.querySelector(`ui5-list`);
+	}
+
+	get _role() {
+		return "tree";
+	}
+
+	_onListItemStepIn(event) {
+		if (this.busy) {
+			return;
+		}
+
+		const listItem = event.detail.item;
+		const treeItem = listItem.treeItem;
+		if (treeItem.items.length > 0) {
+			const firstChildId = treeItem.items[0]._id;
+			const firstChildListItem = this.list.getSlottedNodes("items").find(item => item._id === firstChildId);
+			firstChildListItem && this.list.focusItem(firstChildListItem);
+		}
+	}
+
+	_onListItemStepOut(event) {
+		if (this.busy) {
+			return;
+		}
+
+		const listItem = event.detail.item;
+		const treeItem = listItem.treeItem;
+		if (treeItem.parentElement !== this) {
+			const parentId = treeItem.parentElement._id;
+			const parentListItem = this.list.getSlottedNodes("items").find(item => item._id === parentId);
+			parentListItem && this.list.focusItem(parentListItem);
+		}
+	}
+
+	_onListItemToggle(event) {
+		if (this.busy) {
+			return;
+		}
+
+		const listItem = event.detail.item;
+		const treeItem = listItem.treeItem;
+		const defaultPrevented = !this.fireEvent("itemToggle", { item: treeItem }, true);
+		if (!defaultPrevented) {
+			treeItem.toggle();
+		}
+	}
+
+	_onListItemClick(event) {
+		const listItem = event.detail.item;
+		const treeItem = listItem.treeItem;
+		this.fireEvent("itemClick", { item: treeItem });
+	}
+
+	_onListItemDelete(event) {
+		const listItem = event.detail.item;
+		const treeItem = listItem.treeItem;
+		this.fireEvent("itemDelete", { item: treeItem });
+	}
+
+	_onListSelectionChange(event) {
+		const previouslySelectedItems = event.detail.previouslySelectedItems.map(item => item.treeItem);
+		const selectedItems = event.detail.selectedItems.map(item => item.treeItem);
+
+		previouslySelectedItems.forEach(item => {
+			item.selected = false;
+		});
+		selectedItems.forEach(item => {
+			item.selected = true;
+		});
+
+		this.fireEvent("selectionChange", {
+			previouslySelectedItems,
+			selectedItems,
+		});
+	}
+
+	/**
+	 * Perform Depth-First-Search walk on the tree and run a callback on each node
+	 *
+	 * @public
+	 * @param {function} callback function to execute on each node of the tree with 2 arguments: the node and the level
+	 */
+	walk(callback) {
+		walkTree(this, 1, callback);
+	}
+}
+
+const walkTree = (el, level, callback) => {
```

Comment:
Have you considered walkTree & buildTree to be static methods in the Tree class?

Reply:
Yeah, I thought about it, but I prefer them hidden for the moment